File: plugins/extract-ts/clients.ts
// A call to another service, read off the generated client the adapter holds.
//
// An HTTP peer: the adapter's file imports `createClient` from openapi-fetch
// and the `paths` type from its gen/, and a call `client.GET("/v1/x")` names
// the verb and the route in the code; the document vendored beside gen/ says
// which operation answers there. A gRPC peer: the adapter imports a service
// descriptor from a `*_pb` module and creates a Connect client from it; the
// descriptor's name is the service in the vendored .proto, whose package is
// the proto package, and a call `client.getQuote()` is the rpc in the proto's
// own case. Either way the id is the one the callee's extractor would give.

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { basename, dirname, join } from "node:path";
import { callID, findOperation, readSpec, type Spec } from "./openapi.ts";
import type { Source } from "./source.ts";
import { isCall, isMember, isTemplate, memberName, stringOf, templateShape, walk, type CallExpression, type Node } from "./ast.ts";
import type { Diagnostics } from "./domain.ts";

/** One call the adapter makes, in the catalog's terms. */
export interface RpcHop {
  /** "auth.v1.Sessions/validateSession" or "shop.v1.Pricing/GetQuote" */
  id: string;
  /** What the manifest's peers map is keyed by: the api id or the proto package. */
  pkg: string;
  /** The document or the proto the call was named from. */
  source: string;
}

const HTTP_VERBS = new Set(["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS"]);

export interface ProtoService {
  pkg: string;
  name: string;
  rpcs: string[];
  source: string;
}

/** What one adapter file can reach: its HTTP document and its proto services. */
class Peer {
  readonly spec: Spec | undefined;
  readonly protos: ProtoService[];
  constructor(spec: Spec | undefined, protos: ProtoService[]) {
    this.spec = spec;
    this.protos = protos;
  }
}

const peers = new Map<string, Peer>();

function peerOf(src: Source, rel: (abs: string) => string, b: Diagnostics): Peer {
  const hit = peers.get(src.path);
  if (hit) return hit;

  let spec: Spec | undefined;
  const protos: ProtoService[] = [];
  for (const imp of src.imports) {
    if (!imp.file) continue;
    // gen/types.ts of openapi-typescript → gen/openapi.yaml beside it
    if (/[\\/]gen[\\/]/.test(imp.file) && imp.file.endsWith(".ts") && !/_pb\.ts$/.test(imp.file)) {
      const found = specBeside(dirname(imp.file));
      if (found) spec = readSpec(found);
    }
    // gen/**/<x>_pb.ts of Connect-ES → proto/**/*.proto beside gen/
    if (/_pb\.ts$/.test(imp.file)) {
      const gen = genDirAbove(imp.file);
      if (gen) protos.push(...readProtos(join(dirname(gen), "proto"), rel));
    }
  }
  if (!spec && src.imports.some((i) => i.specifier === "openapi-fetch")) {
    b.warn(rel(src.path), "creates an openapi-fetch client but imports no gen/types beside an openapi document, so its calls cannot be named");
  }
  if (protos.length === 0 && src.imports.some((i) => i.specifier.startsWith("@connectrpc/"))) {
    b.warn(rel(src.path), "creates a Connect client but imports no *_pb descriptor with a proto beside it, so its calls cannot be named");
  }
  const peer = new Peer(spec, protos);
  peers.set(src.path, peer);
  return peer;
}

/** The gen/ a generated file sits under, however deep its package path put it. */
function genDirAbove(file: string): string | undefined {
  let dir = dirname(file);
  for (let i = 0; i < 8 && dir !== dirname(dir); i++) {
    if (basename(dir) === "gen") return dir;
    dir = dirname(dir);
  }
  return undefined;
}

function specBeside(dir: string): string | undefined {
  for (const name of ["openapi.yaml", "openapi.yml", "swagger.yaml", "swagger.yml"]) {
    const path = join(dir, name);
    if (existsSync(path)) return path;
  }
  return undefined;
}

/** A minimal reading of .proto files: package, services, rpcs. */
export function readProtos(dir: string, rel: (abs: string) => string): ProtoService[] {
  if (!existsSync(dir)) return [];
  const out: ProtoService[] = [];
  const walk = (d: string): void => {
    for (const name of readdirSync(d).sort()) {
      const p = join(d, name);
      if (statSync(p).isDirectory()) walk(p);
      else if (name.endsWith(".proto")) out.push(...protoServices(readFileSync(p, "utf8"), rel(p)));
    }
  };
  walk(dir);
  return out;
}

function protoServices(text: string, source: string): ProtoService[] {
  const stripped = text.replace(/\/\/.*$/gm, "").replace(/\/\*[\s\S]*?\*\//g, "");
  const pkg = /^\s*package\s+([\w.]+)\s*;/m.exec(stripped)?.[1] ?? "";
  const out: ProtoService[] = [];
  const service = /\bservice\s+(\w+)\s*\{([^}]*)\}/g;
  for (let m = service.exec(stripped); m; m = service.exec(stripped)) {
    const rpcs = [...m[2]!.matchAll(/\brpc\s+(\w+)\s*\(/g)].map((r) => r[1]!);
    out.push({ pkg, name: m[1]!, rpcs, source });
  }
  return out;
}

/**
 * The calls one method of an adapter makes to its peer, in order. Read off
 * the method's body: `x.GET("/route")` for HTTP, `x.getQuote(...)` for gRPC.
 */
export function adapterCalls(src: Source, cls: string, method: string, rel: (abs: string) => string, b: Diagnostics): RpcHop[] {
  const c = src.classes.find((k) => k.name === cls);
  const m = c?.methods.get(method);
  if (!m?.body) return [];
  const peer = peerOf(src, rel, b);
  const out: RpcHop[] = [];
  const seen = new Set<string>();
  walk(m.body, (node) => {
    if (!isCall(node) || !isMember(node.callee)) return;
    const name = memberName(node.callee);
    const hop = name === undefined ? undefined : hopOf(peer, name, node, rel, b);
    if (hop && !seen.has(hop.id)) {
      seen.add(hop.id);
      out.push(hop);
    }
  });
  return out;
}

/** Calls from any body that reaches a generated client directly, by the same reading. */
export function callsIn(src: Source, node: Node, rel: (abs: string) => string, b: Diagnostics): RpcHop[] {
  const peer = peerOf(src, rel, b);
  if (!peer.spec && peer.protos.length === 0) return [];
  const out: RpcHop[] = [];
  walk(node, (n) => {
    if (!isCall(n) || !isMember(n.callee)) return;
    const name = memberName(n.callee);
    const hop = name === undefined ? undefined : hopOf(peer, name, n, rel, b);
    if (hop) out.push(hop);
  });
  return out;
}

function hopOf(peer: Peer, name: string, call: CallExpression, rel: (abs: string) => string, b: Diagnostics): RpcHop | undefined {
  if (HTTP_VERBS.has(name)) {
    if (!peer.spec) return undefined;
    const first = call.arguments[0];
    if (!first) return undefined;
    const route = stringOf(first) ?? (isTemplate(first) ? templateShape(first) : undefined);
    if (!route) return undefined;
    const op = findOperation(peer.spec, name, route);
    if (!op) {
      b.warn(rel(peer.spec.source), `calls ${name} ${route}, which the document does not declare; the call is left out`);
      return undefined;
    }
    return { id: callID(peer.spec, op), pkg: peer.spec.api, source: rel(peer.spec.source) };
  }
  for (const svc of peer.protos) {
    const rpc = svc.rpcs.find((r) => r.toLowerCase() === name.toLowerCase());
    if (rpc) return { id: `${svc.pkg}.${svc.name}/${rpc}`, pkg: svc.pkg, source: svc.source };
  }
  return undefined;
}

export function isOpenapiFetchClientType(type: string): boolean {
  return /^Client<\s*paths\s*>$/.test(type.trim());
}

File: plugins/extract-django/clients.py
"""A call to another service, read off the client that makes it.

`<app>/clients/<peer>/` holds the vendored document and the class that calls
it. The code names the verb and the route - `self._http.post("/v1/quotes")` -
and the document beside it says which operation answers there, so the call is
recorded under the id the callee's own extractor would give it. Without a
document there is no id to share: the call is still recorded, against the route
it names, and left unresolved.
"""

from __future__ import annotations

import ast
import os
from dataclasses import dataclass, field as dc_field
from typing import Dict, List, Optional, Tuple

import openapi
from apps import App
from source import Module, const_str, dotted

VERBS = {"get", "post", "put", "patch", "delete", "head", "options", "request"}


@dataclass
class Call:
    """One method of a client, and what it reaches."""

    id: str  # "shop.v1.Pricing/GetQuote", or "POST /v1/quotes" when nothing names it
    pkg: str  # what the manifest's peers map is keyed by: the api id
    source: str  # the document, or the client module
    label: str  # what the arrow says: the operation, or the route
    resolved: bool


@dataclass
class Client:
    name: str  # the class
    module: Module
    directory: str
    spec: Optional[openapi.Spec] = None
    calls: Dict[str, Call] = dc_field(default_factory=dict)  # method -> call


def path_of(node: ast.AST) -> str:
    """The route as the code writes it, with every hole spelled the same."""
    literal = const_str(node)
    if literal:
        return literal
    if isinstance(node, ast.JoinedStr):
        out = ""
        for part in node.values:
            if isinstance(part, ast.Constant) and isinstance(part.value, str):
                out += part.value
            else:
                out += "{}"
        return out
    return ""


def http_call(node: ast.Call) -> Optional[Tuple[str, str]]:
    """(verb, route) for a call that goes out over HTTP.

    The route has to look like one. `get` is the name of an HTTP verb and also
    the name of the method every mapping in Python has, and what tells the two
    apart is that only one of them is handed a path."""
    name = dotted(node.func)
    verb = name.split(".")[-1]
    if verb not in VERBS or not node.args:
        return None
    if verb == "request":
        if len(node.args) < 2:
            return None
        route = path_of(node.args[1])
        return (const_str(node.args[0]).upper(), route) if is_route(route) else None
    route = path_of(node.args[0])
    return (verb.upper(), route) if is_route(route) else None


def is_route(route: str) -> bool:
    return route.startswith("/") or route.startswith("http://") or route.startswith("https://")


def read_clients(app: App, peers: Dict[str, str], rel, b) -> List[Client]:
    out: List[Client] = []
    for module in app.package("clients"):
        directory = os.path.dirname(module.path)
        spec_path = openapi.beside(directory)
        spec = openapi.read(spec_path) if spec_path else None
        for node in module.classes():
            client = Client(name=node.name, module=module, directory=directory, spec=spec)
            for method in node.body:
                if not isinstance(method, (ast.FunctionDef, ast.AsyncFunctionDef)):
                    continue
                for child in ast.walk(method):
                    if not isinstance(child, ast.Call):
                        continue
                    found = http_call(child)
                    if found is None:
                        continue
                    verb, route = found
                    client.calls[method.name] = resolve(verb, route, spec, module, os.path.basename(directory), rel, b)
                    break
            if client.calls:
                out.append(client)
        if spec is None and any(module.classes()):
            b.warn(module.rel, "no OpenAPI document beside this client: its calls name a route, and nothing says which operation answers there")
    for client in out:
        for call in client.calls.values():
            if call.resolved and call.pkg not in peers:
                b.warn(call.id, "%s answers this call and the manifest names no peer for it: the step stays unresolved" % call.pkg)
    return out


def resolve(verb: str, route: str, spec: Optional[openapi.Spec], module: Module, fallback: str, rel, b) -> Call:
    if spec is not None:
        operation = spec.find(verb, route)
        if operation is not None:
            return Call(id=operation.call_id(spec.api), pkg=spec.api, source=rel(spec.path), label=operation.id, resolved=True)
        b.warn(module.rel, "%s %s is not a route %s declares" % (verb, route, rel(spec.path)))
    return Call(id="%s %s" % (verb, route), pkg=fallback, source=module.rel, label="%s %s" % (verb, route), resolved=False)
